### PR TITLE
fix refiner switching issues

### PR DESCRIPTION
**Pull request by xhoxye**
https://github.com/lllyasviel/Fooocus/issues/2131

1. In inpaint, when refiner is None, enable use_synthetic_refiner. The default switching timing of 0.5 is too early, which is now modified to SDXL default of 0.8.
2. When using custom steps, the calculation of switching timing is wrong. Now it is modified to calculate "steps x timing" after custom steps are used.

### CODE REVIEW

**Review by mashb1t — APPROVED** [review]

File: modules/async_worker.py [hunk 1]
```diff
@@ -304,7 +304,7 @@ def handler(async_task):
                         print(f'[Inpaint] Current inpaint model is {inpaint_patch_model_path}')
                         if refiner_model_name == 'None':
                             use_synthetic_refiner = True
-                            refiner_switch = 0.5
+                            refiner_switch = 0.8
```
Comment (mashb1t) [inline modules/async_worker.py:307]: @xhoxye can you please elaborate why exactly 0.8 should be the new value here?

Comment (xhoxye) [reply modules/async_worker.py:307]: 0.8 is the default switching timing value for SDXL refiner, which reduces the loss of control of the picture to a small effect, of course 1.0 is OK, but 1.0 will turn off use_synthetic_refiner

Comment (mashb1t) [reply modules/async_worker.py:307]: I'm uncertain what the purpose of using use_synthetic_refiner is. Does it allow for more unguided steps? But this is what the ControlNet stop at is for anyways, right?

Comment (mashb1t) [reply modules/async_worker.py:307]: i've tested both variants, and for me basically nothing changes when setting refiner_switch to 0.5 or 0.8 without an actual refiner. 

Prompt: "a man wearing a red suit"
Seed: 0
Resolution: 1152×896 (default)
Stop At: 0.9 (default)
Weight: 0.75 (default)

Base image:
base

0.5:
image - 2024-02-11T134154 584

0.8:
image - 2024-02-11T140123 006

I'd merge the fix for refiner switch calculation but not for the first part until clarified further.
Is there something i'm missing?

**Comment by xhoxye** [comment]
Can you get another Asian face. Try it.
it becomes a European after 0.5.
QQ截图20240211213714
0.5

**Comment by mashb1t** [comment]
yes, got it. Updated comment https://github.com/lllyasviel/Fooocus/pull/2165#discussion_r1485592011

File: modules/async_worker.py (hunk 1 above)
Comment (xhoxye) [reply modules/async_worker.py:307]: Can you get another Asian face. Try it.
it becomes a European after 0.5 or 0.7.
QQ截图20240211214143

image

**Comment by mashb1t** [comment]
> Could it be that Asians face and models didn't match?

This could be possible, still switching to refiner, even synthetic one, will again introduce noise depending on switch step and cause the result to not apply faceswap / inpaint correctly

**Comment by mashb1t** [comment]
@lllyasviel IMHO this is a non-intended side-effect but it's also possible that we're not getting the full picture here.
Is it intended that refiner_switch is 0.5 and not 1 when no refiner is set and if yes, what are the reasons?

https://github.com/lllyasviel/Fooocus/blob/2.1.864/modules/async_worker.py#L307

**Comment by mashb1t** [comment]
@lllyasviel your feedback would be highly appreciated here to identify a potential bug and to understand the implementation better.

**Comment by ChineseNorris** [comment]
I've checked a few things with the refiner_switch = 0.8 and here are some issues I have found:

- With refiner_switch = 0.8 the image gets more saturated and looks a little bit "better" but the inpaint(masked) creates a thin white line that looks like it's glowing
- With refiner_switch = 0.5 the image looks more natural and the inpaint doesn't do weird things
- With refiner_switch = 1 the saturation is more clear and doesn't look as natural as 0.5

Maybe change refiner_switch to be altered in profiles and change this setting if anyone needs it?
refiner_switch = 0.8

refiner_switch = 0.5

**Pull request merged**